candle-metal-kernels/src/reduce.metal

```diff
@@ -19,288 +21,637 @@ METAL_FUNC uint get_strided_index(
-REDUCE(x * y, fast_mul_f16_strided, half, 1)
-REDUCE(MAX(x, y), fast_max_f32_strided, float, -HUGE_VALF)
-REDUCE(MAX(x, y), fast_max_u32_strided, uint, 0)
-REDUCE(MAX(x, y), fast_max_f16_strided, half, -HUGE_VALH)
-REDUCE(MAX(x, y), fast_max_u8_strided, uint8_t, 0)
-REDUCE(MIN(x, y), fast_min_f32_strided, float, HUGE_VALF)
-REDUCE(MIN(x, y), fast_min_u32_strided, uint, 0xFFFFFFFF)
-REDUCE(MIN(x, y), fast_min_f16_strided, half, HUGE_VALH)
-REDUCE(MIN(x, y), fast_min_u8_strided, uint8_t, 0xFF)
-ARGMIN(fast_argmin_f32_strided, float, HUGE_VALF)
-ARGMIN(fast_argmin_f16_strided, half, HUGE_VALH)
-ARGMIN(fast_argmin_u32_strided, uint, 0xFFFFFFFF)
-ARGMIN(fast_argmin_u8_strided, uint8_t, 0xFF)
-ARGMAX(fast_argmax_f32_strided, float, -HUGE_VALF)
-ARGMAX(fast_argmax_f16_strided, half, -HUGE_VALH)
-ARGMAX(fast_argmax_u32_strided, uint, 0)
-ARGMAX(fast_argmax_u8_strided, uint8_t, 0)
-
-SOFTMAX(softmax_f32, float)
-SOFTMAX(softmax_f16, half)
+template <typename V>
+struct Indexed {
+    uint i;
+    V val;
+    typedef V type;
+
+    constexpr Indexed<V>() thread = default;
+    constexpr Indexed<V>() threadgroup = default;
+    constexpr Indexed<V>() device = default;
+    constexpr Indexed<V>() constant = default;
+
+    constexpr Indexed<V>(uint _i, V _val) : i(_i), val(_val) {}
+
+    template <typename U, typename = typename enable_if<is_convertible_v<U, V>>::type>
+    constexpr Indexed<V>(uint _i, U _val) : i(_i), val(static_cast<U>(_val)) {}
+
+    template <typename U>
+    constexpr Indexed<V>(const thread Indexed<U> &iv): Indexed<V>(iv.i, iv.val) {}
+
+    template <typename U>
+    constexpr Indexed<V>(const threadgroup Indexed<V> &iv): Indexed<V>(iv.i, iv.val) {}
+
+    Indexed<V> operator=(const thread Indexed<V> &iv) thread {
+        this->i = iv.i;
+        this->val = iv.val;
+        return *this;
+    }
+    Indexed<V> operator=(const thread Indexed<V> &iv) threadgroup {
+        this->i = iv.i;
+        this->val = iv.val;
+        return *this;
+    }
+};
+
+template<typename V>
+constexpr METAL_FUNC bool operator<(Indexed<V> lhs, Indexed<V> rhs) {
+    return lhs.val < rhs.val || (lhs.val == rhs.val && lhs.i < rhs.i);
+}
+
+template<typename V>
+constexpr METAL_FUNC bool operator>(Indexed<V> lhs, Indexed<V> rhs) {
+    return lhs.val > rhs.val || (lhs.val == rhs.val && lhs.i > rhs.i);
+}
+
+template<typename T>
+struct _numeric_limits_impl<Indexed<T>> {
+    static constexpr Indexed<T> lowest() {
+        return Indexed<T>(0, numeric_limits<T>::lowest());
+    }
+
+    static constexpr Indexed<T> max() {
+        return Indexed<T>(0, numeric_limits<T>::max());
+    }
+};
+
+#if defined(__HAVE_BFLOAT__)
+// Metal does not have simd_shuffle_down for bfloat16
+// TODO: Check if volatile threadgroup memory reduction is faster than simd_shuffle_down for bfloat
+bfloat simd_shuffle_down(bfloat value, ushort delta) {
+    return static_cast<bfloat>(__metal_simd_shuffle_down(static_cast<float>(value), delta));
```

Comment:
I'm surprised they don't have `simd_shuffle_down` for bf16.

Reply:
myself & @ivarflakstad dug into it - they don't!

Reply:
And bfloat -> float simd shuffle -> bfloat is faster than the alternative method 🙃